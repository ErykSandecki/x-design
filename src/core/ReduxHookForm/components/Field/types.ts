// types
import { TUseInputProps } from './hooks/useInputProps';
import { TUseMetaProps } from './hooks/useMetaProps';

export type TFieldRestProps<T, V> = TUseInputProps<V> &
  T & {
    ref?: any;
  };

export type TFieldComponentProps<T, V> = Partial<TUseMetaProps<V>> & TFieldRestProps<T, V>;
