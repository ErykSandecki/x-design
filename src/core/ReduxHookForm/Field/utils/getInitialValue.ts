import { head, isArray } from 'lodash';

// types
import { TFieldValue } from '../../types';

export const getComplexInitalValue = <V>(initValue: Array<boolean | string>, shouldBeEmpty: boolean): V => {
  const length = initValue.length;
  const isBoolean = typeof head(initValue) === 'boolean';
  const values = Array.from(Array(length), () => (isBoolean ? false : ''));

  return shouldBeEmpty ? (values as V) : (initValue as V);
};

export const getInitialValue = <V extends TFieldValue>(
  defaultValue: V,
  formatOnInit: (value: V) => V,
  shouldBeEmpty: boolean,
): V => {
  const initValue = formatOnInit ? formatOnInit((defaultValue as V) || ('' as V)) : (defaultValue as V);

  switch (true) {
    case typeof initValue === 'boolean':
      return shouldBeEmpty ? (false as V) : initValue;
    case typeof initValue === 'number':
      return shouldBeEmpty ? (0 as V) : initValue;
    case typeof initValue === 'string':
      return shouldBeEmpty ? ('' as V) : initValue;
    case isArray(initValue) && !!initValue.length:
      return getComplexInitalValue<V>(initValue, shouldBeEmpty);
    default:
      return null;
  }
};
