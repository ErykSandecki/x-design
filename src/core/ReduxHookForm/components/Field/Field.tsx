import {
  FC,
  ForwardRefExoticComponent,
  ReactNode,
  RefObject,
  useContext,
  useEffect,
} from 'react';
import { isEmpty } from 'lodash';
import { useSelector } from 'react-redux';

// hooks
import { useMetaProps } from './hooks/useMetaProps';
import { useInitField } from './hooks/useInitField';
import { useInputProps } from './hooks/useInputProps';

// others
import { FormContext } from '../Form/Form';

// store
import { fieldSelectorCreator } from 'store/reduxHookForm/selectors';

// types
import { TAsyncValidator, TFieldValue, TSyncValidator } from '../../types';
import { TFieldComponentProps } from './types';
import { TObject } from 'types';

export type TFieldProps<T, V> = {
  afterSubmit?: () => void;
  asyncValidators?: Array<TAsyncValidator>;
  beforeSubmit?: () => void;
  Component:
    | FC<TFieldComponentProps<T, V>>
    | ForwardRefExoticComponent<TFieldComponentProps<T, V>>;
  data?: TObject<any>;
  defaultValue?: TFieldValue;
  fieldsToClearOnChange?: Array<string>;
  formatOnBlur?: (value: V, name: string) => V;
  formatOnChange?: (value: V, name: string) => V;
  formatOnFocus?: (value: V, name: string) => V;
  formatOnInit?: (value: V) => V;
  name: string;
  parse?: (value: V, name: string) => V;
  ref?: RefObject<HTMLElement>;
  subscriptionFields?: Array<string>;
  syncValidators?: Array<TSyncValidator>;
  touched?: boolean;
  visited?: boolean;
} & T;

export const Field = <T, V extends TFieldValue>({
  afterSubmit,
  asyncValidators = [],
  beforeSubmit,
  Component,
  data,
  defaultValue,
  fieldsToClearOnChange = [],
  formatOnBlur,
  formatOnChange,
  formatOnFocus,
  formatOnInit,
  name,
  parse,
  ref,
  subscriptionFields = [],
  syncValidators = [],
  touched: initialTouched = false,
  visited: initialVisited = false,
  ...restProps
}: TFieldProps<T, V>): ReactNode => {
  const formName = useContext(FormContext);
  const field = useSelector(fieldSelectorCreator(formName, name));
  const inputProps = useInputProps<V>(formName, name);
  const metaProps = useMetaProps<V>(formName, name);
  const { touched, value = '', visited } = field;

  const {
    subscriptionFieldsValues,
    updateAsyncValidators,
    updateSyncValidators,
  } = useInitField(
    afterSubmit,
    asyncValidators,
    beforeSubmit,
    data,
    defaultValue,
    fieldsToClearOnChange,
    formatOnBlur,
    formatOnChange,
    formatOnFocus,
    formatOnInit,
    name,
    parse,
    subscriptionFields,
    syncValidators,
    initialTouched,
    initialVisited,
  );

  useEffect(() => {
    if (!isEmpty(field)) {
      updateAsyncValidators(value);
      updateSyncValidators(value);
    }
  }, [touched, value, visited, ...subscriptionFieldsValues]);

  return (
    <Component
      meta={metaProps}
      {...inputProps}
      {...(restProps as T)}
      {...(ref ? { ref: ref } : {})}
    />
  );
};

export default Field;
