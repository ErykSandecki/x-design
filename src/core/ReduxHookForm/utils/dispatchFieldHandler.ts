import { Dispatch } from 'redux';

// types
import { TField, TUpdateFieldActions } from 'store/reduxHookForm/types';

// store
import { updateField } from 'store/reduxHookForm/actions';

export const dispatchFieldHandler =
  (dispatch: Dispatch, formName: string, name: string) =>
  (field: Partial<TField>, actionType: TUpdateFieldActions): void => {
    dispatch(
      updateField(
        {
          field: {
            ...field,
          },
          formName,
          name,
        },
        actionType,
      ),
    );
  };
