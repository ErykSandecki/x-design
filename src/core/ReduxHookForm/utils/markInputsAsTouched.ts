import { Dispatch } from 'redux';
import { keys } from 'lodash';

// others
import { SET_TOUCHED_FIELD } from 'store/reduxHookForm/actionsType';

// store
import { updateField } from 'store/reduxHookForm/actions';

// types
import { TFields } from 'store/reduxHookForm/types';

export const markInputsAsTouched = (dispatch: Dispatch, formName: string, fields: TFields): void => {
  const fieldNames = keys(fields);

  fieldNames.forEach((fieldName) => {
    dispatch(
      updateField(
        {
          field: { touched: true },
          formName,
          name: fieldName,
        },
        SET_TOUCHED_FIELD,
      ),
    );
  });
};
