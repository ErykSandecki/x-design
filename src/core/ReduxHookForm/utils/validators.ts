import { every } from 'lodash';

// types
import { TAsyncValidator, TFieldValue, TSyncValidator } from '../types';
import { TFields } from 'store/reduxHookForm/types';
import { TFileData } from 'types/types';

export const defaultFormValidator = (fields: TFields): boolean =>
  every(fields, (field) => !(field.asyncErrors.length || field.syncErrors.length));

export const formValidatorWithFileUploader =
  (files: Array<TFileData>, setTouched: TFunc<[boolean]>) =>
  (fields: TFields): boolean => {
    setTouched(true);

    return files.length !== 0 && every(fields, (field) => !(field.asyncErrors.length || field.syncErrors.length));
  };

export const getErrorsFromAsyncValidators = async (
  validators: Array<TAsyncValidator>,
  value: TFieldValue,
  subscribedFields: TFields | undefined,
  t: TT,
): Promise<Array<string>> => {
  const errors = [];

  for (const validator of validators) {
    errors.push(await validator(t, value, subscribedFields));
  }

  return errors.filter(Boolean);
};

export const getErrorsFromSyncValidators = (
  validators: Array<TSyncValidator>,
  value: TFieldValue,
  subscribedFields: TFields | undefined,
  t: TT,
): Array<string> => validators.map((validator) => validator(t, value, subscribedFields)).filter(Boolean);
