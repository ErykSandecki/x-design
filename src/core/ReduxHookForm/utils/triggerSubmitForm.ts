import { Dispatch } from 'redux';

// others
import { UPDATE_FORM_VALIDATOR } from 'store/reduxHookForm/actionsType';

// store
import { submit, updateForm } from 'store/reduxHookForm/actions';

// types
import { TFields } from 'store/reduxHookForm/types';

// utils
import { markInputsAsTouched } from './markInputsAsTouched';

export const triggerSubmitForm = (dispatch: Dispatch, fields: TFields, formName: string, isValid: boolean): void => {
  markInputsAsTouched(dispatch, formName, fields);
  dispatch(submit(formName));
  dispatch(updateForm({ form: { isValid }, formName }, UPDATE_FORM_VALIDATOR));
};
