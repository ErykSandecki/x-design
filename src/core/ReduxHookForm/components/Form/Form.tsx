import { createContext, FormEvent, ReactNode, useEffect } from 'react';
import { forOwn, isEmpty, kebabCase } from 'lodash';
import { useDispatch, useSelector } from 'react-redux';

// others
import { UPDATE_FORM_VALIDATOR } from '../../../../store/reduxHookForm/actionsType';

// store
import {
  destroyForm,
  mountForm,
  submit,
  updateForm,
} from '../../../../store/reduxHookForm/actions';
import {
  fieldsSelectorCreator,
  formSelectorCreator,
} from '../../../../store/reduxHookForm/selectors';

// types
import { E2EAttribute } from 'types/e2e';
import {
  E2EDataAttribute,
  TE2EDataAttributeProps,
} from 'shared/E2EDataAttributes/E2EDataAttribute';
import { TField, TFields } from '../../../../store/reduxHookForm/types';

// utils
import { defaultFormValidator } from '../../utils/validators';
import { markInputsAsTouched } from '../../utils/markInputsAsTouched';

export const FormContext = createContext('');

export type TFormProps<T> = {
  asyncTimeDelay?: number;
  children: ReactNode;
  className?: string;
  e2eAttribute?: TE2EDataAttributeProps['type'];
  e2eValue?: TE2EDataAttributeProps['value'];
  formName: string;
  isValid?: boolean;
  onSubmit: (formData: T) => void;
  validate?: (fields: TFields<T>) => boolean;
};

export const Form = <T extends {}>({
  asyncTimeDelay = 0,
  children,
  className = '',
  e2eAttribute = E2EAttribute.form,
  e2eValue = '',
  formName,
  isValid: isValidInitial = false,
  onSubmit,
  validate = defaultFormValidator,
}: TFormProps<T>): ReactNode => {
  const dispatch = useDispatch();
  const form = useSelector(formSelectorCreator(formName));
  const fields = useSelector(fieldsSelectorCreator(formName));

  const getFieldsValues = (): T => {
    const formData = {};

    forOwn(fields, (field: TField, name: string) => {
      const { parse, value } = field;

      formData[name] = parse ? parse(value, name) : value;
    });

    return formData as T;
  };

  const onValidateHandler = (): void => {
    const isValid = validate(fields as TFields<T>);

    if (isValid) {
      onSubmit(getFieldsValues());
    }

    dispatch(
      updateForm({ form: { isValid }, formName }, UPDATE_FORM_VALIDATOR),
    );
  };

  const onSubmitHandler = (event: FormEvent): void => {
    event.preventDefault();

    dispatch(submit(formName));
    markInputsAsTouched(dispatch, formName, fields);
    onValidateHandler();
  };

  useEffect(() => {
    if (isEmpty(form)) {
      dispatch(
        mountForm({
          [formName]: {
            asyncTimeDelay,
            error: '',
            fields: {},
            isPending: false,
            isValid: isValidInitial,
          },
        }),
      );
    }

    return (): void => {
      dispatch(destroyForm(formName));
    };
  }, []);

  return (
    <E2EDataAttribute
      type={e2eAttribute}
      value={
        e2eValue ? `${kebabCase(formName)}-${e2eValue}` : kebabCase(formName)
      }
    >
      <form className={className} onSubmit={onSubmitHandler}>
        <FormContext.Provider value={formName}>{children}</FormContext.Provider>
      </form>
    </E2EDataAttribute>
  );
};

export default Form;
