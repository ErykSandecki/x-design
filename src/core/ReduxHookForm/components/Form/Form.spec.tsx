import { fireEvent, render } from '@testing-library/react';
import { Provider } from 'react-redux';

// components
import Field from '../Field/Field';
import Form from './Form';

// mocks
import { reduxHookFormStateMock } from 'test/mocks/reducer/reduxHookFormMock';

// others
import { REDUCER_KEY as REDUX_HOOK_FORM } from 'store/reduxHookForm/actionsType';

// store
import { configureStore } from 'store/store';

// utils
import { required } from 'utils';

const stateMock = {
  ...reduxHookFormStateMock,
};

const Button = (): any => {
  return <button type="submit">Click</button>;
};

const TextField = (): any => {
  return <input type="text" />;
};

describe('Form snapshots', () => {
  it('should render Form', () => {
    // mock
    const store = configureStore(stateMock);

    // before
    const { asFragment, getByText } = render(
      <Provider store={store}>
        <Form onSubmit={() => {}} formName="testForm">
          <Field Component={TextField} name="testField" />
          <Button />
        </Form>
      </Provider>,
    );

    // action
    fireEvent.click(getByText('Click'));

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with passed e2evalue', () => {
    // mock
    const store = configureStore(stateMock);

    // before
    const { asFragment, getByText } = render(
      <Provider store={store}>
        <Form e2eValue="e2eValue" onSubmit={() => {}} formName="testForm">
          <Field Component={TextField} name="testField" />
          <Button />
        </Form>
      </Provider>,
    );

    // action
    fireEvent.click(getByText('Click'));

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render Form when is not valid', () => {
    // mock
    const store = configureStore(stateMock);

    // before
    const { asFragment, getByText } = render(
      <Provider store={store}>
        <Form onSubmit={() => {}} formName="testForm">
          <Field Component={TextField} name="testField" parse={(value: string) => value} syncValidators={[required]} />
          <Button />
        </Form>
      </Provider>,
    );

    // action
    fireEvent.click(getByText('Click'));

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with init form', () => {
    // mock
    const store = configureStore({
      [REDUX_HOOK_FORM]: {},
    });

    // before
    const { asFragment, getByText } = render(
      <Provider store={store}>
        <Form onSubmit={() => {}} formName="testForm">
          <Field Component={TextField} name="testField" parse={(value: string) => value} syncValidators={[required]} />
          <Button />
        </Form>
      </Provider>,
    );

    // action
    fireEvent.click(getByText('Click'));

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render Form with not parsed value', () => {
    // mock
    const store = configureStore({
      [REDUX_HOOK_FORM]: {
        ...reduxHookFormStateMock[REDUX_HOOK_FORM],
        testForm: {
          ...reduxHookFormStateMock[REDUX_HOOK_FORM].testForm,
          fields: {
            ...reduxHookFormStateMock[REDUX_HOOK_FORM].testForm.fields,
            testField: {
              ...reduxHookFormStateMock[REDUX_HOOK_FORM].testForm.fields.testField,
              parse: undefined,
            },
          },
        },
      },
    });

    // before
    const { asFragment, getByText } = render(
      <Provider store={store}>
        <Form onSubmit={() => {}} formName="testForm">
          <Field Component={TextField} name="testField" parse={(value: string) => value} />
          <Button />
        </Form>
      </Provider>,
    );

    // action
    fireEvent.click(getByText('Click'));

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
