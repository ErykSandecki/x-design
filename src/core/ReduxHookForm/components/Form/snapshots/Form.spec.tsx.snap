// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form snapshots should render Form 1`] = `
<DocumentFragment>
  <form
    class=""
    data-test-form="test-form"
  >
    <input
      type="text"
    />
    <button
      type="submit"
    >
      Click
    </button>
  </form>
</DocumentFragment>
`;

exports[`Form snapshots should render Form when is not valid 1`] = `
<DocumentFragment>
  <form
    class=""
    data-test-form="test-form"
  >
    <input
      type="text"
    />
    <button
      type="submit"
    >
      Click
    </button>
  </form>
</DocumentFragment>
`;

exports[`Form snapshots should render Form with not parsed value 1`] = `
<DocumentFragment>
  <form
    class=""
    data-test-form="test-form"
  >
    <input
      type="text"
    />
    <button
      type="submit"
    >
      Click
    </button>
  </form>
</DocumentFragment>
`;

exports[`Form snapshots should render with init form 1`] = `
<DocumentFragment>
  <form
    class=""
    data-test-form="test-form"
  >
    <input
      type="text"
    />
    <button
      type="submit"
    >
      Click
    </button>
  </form>
</DocumentFragment>
`;

exports[`Form snapshots should render with passed e2evalue 1`] = `
<DocumentFragment>
  <form
    class=""
    data-test-form="test-form-e2eValue"
  >
    <input
      type="text"
    />
    <button
      type="submit"
    >
      Click
    </button>
  </form>
</DocumentFragment>
`;
